pennylane/math/ode.py

```diff
@@ -0,0 +1,246 @@
+        y0 = jnp.eye(2**2, dtype=complex)
+        params = jnp.ones(1, dtype=complex)
+
+        def func(y, t, params):
+            H = jnp.array([XX + params[0] * jnp.sin(t) * ZY])
+            return -1j * jnp.sum(H, axis=0) @ y
+
+        U = qml.math.odeint(func, y0, ts, params)
+
+    Formally, this solution can be written as the time-ordered exponential
+    :math:`U(t_0, t_1) = \text{Texp}\left[-i \int_{t_0}^{t_1}d\tau H(\tau)\right]`
+    (see `Dyson Series <https://en.wikipedia.org/wiki/Dyson_series>`_).
+
+    Note that this computation can be backward differentiated with respect to the parameters ``params`` via
+
+    >>> jac = jax.jacobian(qml.math.odeint, argnums=3, holomorphic=True)(func, y0, ts, params)
+    """
+    y0, unravel = ravel_pytree(y0)
+    func = ravel_first_arg(func, unravel)
+    out = _odeint(func, y0, ts, atol, rtol, *args)
+    return unravel(out)
+
+
+def _abs2(x):
+    """Computing the squared absolute value of x"""
+    if jnp.iscomplexobj(x):
+        return x.real**2 + x.imag**2
+    return x**2
+
+
+def _tolerance_warn(arg, _):
+    # The second blank argument is to abide by host_callback.id_tap's logic
+    atol, rtol, mean_err_ratio, y1_error, err_tol = arg
+    if mean_err_ratio > 1.0:
+        warnings.warn(
+            f"A mean error of {y1_error} in y occured which exceeds the mean error tolerance {err_tol} "
+            f"based on a tolerance of atol = {atol} and rtol = {rtol}. "
+            "Try reducing the step size.",
+            UserWarning,
+        )
+
+
+def _odeint(func, y0, ts, atol, rtol, *args):
+    def func_(y, t):
+        return func(y, t, *args)
+
+    def scan_fun(carry, t1):
+        y0, f0, t0 = carry
+        dt = t1 - t0
+        # not using y1_error and k atm
+        y1, f1, y1_error = runge_kutta_step(func_, y0, f0, t0, dt)
+
+        # check error
+        # def mean_error_ratio(error_estimate, rtol, atol, y0, y1):
+        err_tol = atol + rtol * jnp.maximum(jnp.abs(y0), jnp.abs(y1))
+        err_ratio = y1_error / err_tol.astype(y1_error.dtype)
+        mean_err_ratio = jnp.sqrt(jnp.mean(_abs2(err_ratio)))
+
+        # sends call from device to host to inspect runtime values
+        host_callback.id_tap(
+            _tolerance_warn,
+            (
+                atol,
+                rtol,
+                mean_err_ratio,
+                jnp.sqrt(jnp.mean(_abs2(y1_error))),
+                jnp.sqrt(jnp.mean(_abs2(err_tol))),
+            ),
+        )
+
+        carry = [y1, f1, t1]
+        return carry, y1
+
+    # scan = fancy for-loop in jax, see https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html
+    f0 = func_(y0, ts[0])
+    init = [y0, f0, ts[0]]
+    carry, _ = jax.lax.scan(scan_fun, init, ts[1:])
+    y, _, _ = carry
+    # we typically only care about the final U
+    # In case we want to return all ys, the output of
+    # odeint() must be vmapped, jax.vmap(unravel)(out)
+    return y  # jnp.concatenate((y0[None], ys))
```

Comment:
What is the commented out part for? Did you keep it intentionally?

Reply:
Yes so here I was actually unsure. Typically ode solvers return the full list of evaluated intermediate points (you are returning a "function" in the sense of evaluated points). However, in the context of the pulse programming we dont really care about the intermediate points and just about the final point, which it is currently returning.

Reply:
I like the `kwarg` option, though it would most likely require some sort of if statement here.

There are hacky solutions like `map(partial(jnp.where, condition), return_if_true, return_if_false)` (see e.g. https://github.com/google/jax/blob/main/jax/experimental/ode.py#L208 ) but I cant get it to work

Reply:
Oh dear, ok. Maybe then this should not be high up on the list of priorities :)